src/rgw/rgw_rest_s3.cc

```diff
@@ -6224,18 +6224,33 @@ int RGWSelectObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t ofs, off_
     dump_errno(s);
   }
 
+  auto bl_len = [](bufferlist& bl){int i=0;for(auto& iter : bl.buffers())i++; return i;};
+
   // Explicitly use chunked transfer encoding so that we can stream the result
   // to the user without having to wait for the full length of it.
   if (chunk_number == 0) {
     end_header(s, this, "application/xml", CHUNKED_TRANSFER_ENCODING);
   }
 
   int status=0;
+  int i=0;
+
   for(auto& it : bl.buffers()) {
+
+    ldout(s->cct, 10) << "processing segment " << i << "out of " << bl_len(bl) << " off " << ofs
+                      << " len " << len << " obj-size " << s->obj_size << dendl;
+
+    if(it.length() == 0)
+    {
+      ldout(s->cct, 0) << "s3select:it->_len is zero. segment " << i << " out of " << bl_len(bl)
+                        <<  " obj-size " << s->obj_size << dendl;
+    }
```

Comment:
we should avoid calling `run_s3select()` on this empty segment right?

Reply:
well, that is the issue we had discussed 
/it.length() == 0/ should not happened , not avoiding that will trigger the assert it[0] (accessing beyond ..)
i removed the skipping , in order **not** to bypass this issue.
the log message will point on location

Reply:
i don't agree that this should cause a crash. `bufferlist` is a list of buffer segments, and nothing prevents those segments from being empty. when accessing this memory, you need to do bounds checking on byte [0] the same way you'd have to do it on byte [1]. it's fine if `run_s3select()` has an extra invariant that the input must be non-empty; that just means that calling code has to guarantee that by checking

if the issue is just getting its starting address with `&(it)[0]`, you can use `it.c_str()` instead

Reply:
no crash should happen for zero size segment.
could change the code to access it.c_str() instead of it[0] (and avoid assert)
but,per discussion with @mattbenjamin , zero-segments should be explained (its not the "common case").
in the case zero-segment is valid, will make the change.  

(on previous PR there was a code which skipped the zero-segment, hence no crash)